### PR TITLE
NFV Frontrunning Issue Fix

### LINKED ISSUE #277: Frontrunning Token Transfer
Currently, a NFV transfer TX can be frontrun, where collateral is drained or debt is maxed before transfer, resulting in the loss of value for the buyer.

There are 2 solutions:
1. Use TX nonce or hash of token state with a condition on token transfer to prevent transfer if state change has occurred.
2. Put delay on token transfer, where transfers can only happen after `n` time, so that any change in the amount of debt and/or collateral associated with the token triggers a time delay preventing a token transfer until time is expired.

Solution 1 is preferred, but is incompatible with existing NFT markets. Solution 2 is compatible with existing NFT markets, however it is inconvenient due to `n` time delay.

Expected outcome:
Create a hybrid solution where token defaults to solution 2 for generic NFT markets, however can opt for solution 1 is the NFT market address is registered with Vault721 - meaning it implements a state check using the nonce or state hash of the NFV.

### PR DESCRIPTION
**Pull request by 0xdavinchee**
closes #277 

See [here](https://github.com/open-dollar/od-contracts/issues/277#issuecomment-1867932504) for the implementation details

### CODE REVIEW

File: src/contracts/proxies/Vault721.sol
Comment (daopunk) [inline src/contracts/proxies/Vault721.sol:43]: can you change "_whitelist" to "_allowlist" to avoid any negative connotations

**Comment by daopunk** [comment]
Code looks good. Waiting on feedback from HAI team before approval.

File: src/contracts/proxies/Vault721.sol
Comment (wei3erHase) [inline src/contracts/proxies/Vault721.sol:155]: seems unnecessary to be casting it and storing it as a `uint256`

Comment (wei3erHase) [reply src/contracts/proxies/Vault721.sol:155]: can also do:
```solidity
_hashState[_vaultId] = HashState({
    lastHash: ...,
    lastBlockNumber: ...,
    ...
})
```

File: src/contracts/proxies/Vault721.sol
Comment (wei3erHase) [inline src/contracts/proxies/Vault721.sol:258]: i believe that just calling it `tokenId` (since we're not using the batch transfer) could make it much more readable

File: src/contracts/proxies/Vault721.sol
Comment (wei3erHase) [inline src/contracts/proxies/Vault721.sol:267]: i'm not fully understanding this:
if whitelisted, it reverts with either:
- block delay not passed
- hash different than lastHash

if NOT whitelisted, it only reverts with:
- time delay not passed

why is this difference? what's the point of "whitelisting" an address, to my understanding is:
"this marketplace is whitelisted, so we check the hash on their transfers"
"this marketplace is NOT whitelisted, so we only check timestamp has passed"

Comment (daopunk) [reply src/contracts/proxies/Vault721.sol:267]: so the previous hash and current hash are always compared. however, on non-allowlisted markets, a time-delay of n time is enforced before transferFrom can execute, but on allowlisted markets, a block-delay of 1 or more blocks is enforced because hashes are not only compared internally, but also with the market that checks the hash in the JSON description to ensure that the frontend display of the NFV matches the previous and current hash.

does that clarify it?

**Comment by daopunk** [comment]
@0xdavinchee also, the afterTransfer hook affects both transfer and transferFrom, since it is executed in the internal transfer. however, this problem only really applies to transferFrom, so can you move all the frontrunning-protection logic into a custom hook just for transferFrom, and allow the owner to call transfer without any of these checks

example: the owner wants to move their NFV to another wallet, they should be able to transfer (they will not frontrun themselves) as transfer is not used for market txs

File: src/contracts/proxies/Vault721.sol
Comment (daopunk) [reply src/contracts/proxies/Vault721.sol:267]: basically, allowlisted markets don't have a noticeable delay since the block times on Arbitrum are about 5 seconds or so, which is good for UX

**Comment by 0xdavinchee** [comment]
> @0xdavinchee also, the afterTransfer hook affects both transfer and transferFrom, since it is executed in the internal transfer. however, this problem only really applies to transferFrom, so can you move all the frontrunning-protection logic into a custom hook just for transferFrom, and allow the owner to call transfer without any of these checks
> 
> example: the owner wants to move their NFV to another wallet, they should be able to transfer (they will not frontrun themselves) as transfer is not used for market txs

ah true, I think I will just override `transferFrom` and do the checks in the overridden function in this case as the OZ `transferFrom` function calls `_transfer` which contains the before and after hooks.

File: src/contracts/proxies/Vault721.sol
Comment (pi0neerpat) [inline src/contracts/proxies/Vault721.sol:137]: Magical 🪄 🌈

File: src/contracts/proxies/Vault721.sol
Comment (pi0neerpat) [inline src/contracts/proxies/Vault721.sol:275]: thank you for making this cosmetic change

File: src/contracts/proxies/NFTRenderer.sol
Comment (pi0neerpat) [inline src/contracts/proxies/NFTRenderer.sol:161]: Excellent decision to simplify and remove duplicate code. However, the NFTRenderer was intentionally isolated from the rest of the 721 contract, in order to reduce the security implications of making upgrades. I believe we should continue to keep it limited to only SVG image functionality.

I am marking approved, and we can continue the discussion in #328 for isolating parts of the state-hash feature to its own contract.

**Review by pi0neerpat — APPROVED** [review]

**Pull request merged**
